### PR TITLE
Merge untracked progress on branch folder_project_inventory to 2.0-dev

### PR DESCRIPTION
**Pull request by jiyunyao**
Progress included:
Fixes for deployments
Fix docker tests and precommit hook
Fix proxy binary check
Update gcloud commands in iam_auto_deployment.sh
Refactor the fix on inventory when APIs are disabled for some projects
Fix the inventory when no group in GSuite is created
Test fix when port number is not released in time in case of SQLite

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/forseti-security/839)

### CODE REVIEW

File: google/cloud/forseti/services/inventory/crawler.py [hunk 1]
```diff
@@ -91,24 +91,20 @@ def get_client(self):
...
-        if not resource._warning:
-            resource._warning = warning_message
-        else:
-            resource._warning += warning_message
         self.config.storage.warning(warning_message)
         self.config.progresser.on_warning(error)
 
     def update(self, resource):
-        """Update the error message in the row of the resource
+        """Update the row of an existed resource
```
Comment (blueandgold) [inline google/cloud/forseti/services/inventory/crawler.py]: nit: do you mean ```existing``` rather than ```existed```?

Comment (ahoying) [reply google/cloud/forseti/services/inventory/crawler.py]: Can those files be left out of this PR and merged in post #836? In general I don't think there will be conflicts with my changes, but I'd like to understand the changes in context, and not in with the other unrelated changes in this PR.

Comment (jiyunyao) [reply google/cloud/forseti/services/inventory/crawler.py]: Thanks, I just checked dictionary. I mean "existing", let me fix this

Comment (jiyunyao) [reply google/cloud/forseti/services/inventory/crawler.py]: Sure. We can do that. I don't think there will be any conflicts though.

File: scripts/iam_auto_deployment.sh [hunk 2]
```diff
@@ -13,20 +13,58 @@
...
+BRANCHNAME=""
+SQLINSTANCE=""
+PROJECT_ID=""
+DEPLOYMENTNAME=""
+
+# WARNING: THESE VALUES ARE NOT CONFIGURABLE!
+# These defaults match values contained in the example deployment YAMLs.
+# These should not be considered to be configurable but they are placed
+# here to avoid repetition.
+FORSETI_DB_NAME="forseti_security"
```
Comment (ahoying) [inline scripts/iam_auto_deployment.sh]: This variable is no longer used. Now that we have combined the databases, I recommend just leaving in FORSETI_DB_NAME and replacing all uses of EXPLAIN_DB_NAME with FORSETI_DB_NAME.

File: scripts/iam_auto_deployment.sh [hunk 3]
```diff
@@ -331,25 +368,18 @@ then
 	$CmdPrefix add-iam-policy-binding $ROOTID \
 	 --member=serviceAccount:$SCRAPINGSA \
 	 --role=roles/storage.admin
-	
+
+	gcloud projects add-iam-policy-binding $PROJECT_ID \
```
Comment (ahoying) [inline scripts/iam_auto_deployment.sh]: You also need to add --role=roles/logging.logWriter to the project IAM bindings.

File: deployment-templates/compute-engine/explain-instance.py [hunk 4]
```diff
@@ -46,6 +48,18 @@ def GenerateConfig(context):
 
     resources = []
 
+    EXPORT_INITIALIZE_VARS = (
+        'export SQL_PORT={0}\n'
+        'export SQL_INSTANCE_CONN_STRING="{1}"\n'
+        'export EXPLAIN_DB_NAME="{2}"\n'
```
Comment (ahoying) [inline deployment-templates/compute-engine/explain-instance.py]: I recommend replacing the use of EXPLAIN_DB with FORSETI_DB.

File: google/cloud/forseti/services/inventory/base/resources.py [hunk 5]
```diff
@@ -107,6 +106,16 @@ def parent(self):
     def key(self):
         raise NotImplementedError('Class: {}'.format(self.__class__.__name__))
 
+    def add_warning(self, warning):
+        warning_message = '{}\n'.format(warning)
```
Comment (ahoying) [inline google/cloud/forseti/services/inventory/base/resources.py]: What about making self._warning a list that you append to, and then in get_warning() return '\n'.join(self._warning) ? Then you don't need to build a new string for each new warning.

File: google/cloud/forseti/services/inventory/storage.py [hunk 6]
```diff
@@ -571,25 +606,24 @@ def update(self, resource):
...
+
+        try:
+            new_rows = Inventory.from_resource(self.index, resource)
+            old_rows = self._get_resource_rows(resource.key())
+
+            new_dict = {row.type_class : row for row in new_rows}
+            old_dict = {row.type_class : row for row in old_rows}
+
+            for type_class in InventoryTypeClass.SUPPORTED_TYPECLASS:
+                if type_class in new_dict:
```
Comment (ahoying) [inline google/cloud/forseti/services/inventory/storage.py]: Is there ever a case where there are rows in old_dict that are not in new_dict that should be deleted?

File: scripts/iam_auto_deployment.sh [hunk 7]
```diff
@@ -376,13 +406,10 @@ then
...
 	BRANCHNAME=$DownloadBranch
 fi
-sed -i -e 's/BRANCHNAME/'$BRANCHNAME'/g' \
-$repodir/deployment-templates/deploy-explain.yaml
 
 # sql instance name
 echo -e "${TYELLOW}Choosing SQL Instance name${TNC}"
-timestamp=$(date --utc +%Ft%Tz | sed -e 's/:/-/g')
-SQLINSTANCE="iam-explain-no-external-"$timestamp
+SQLINSTANCE="iam-explain-no-external-"$LABEL_SAFE_TIMESTAMP
```
Comment (ahoying) [inline scripts/iam_auto_deployment.sh]: replace iam-explain reference(s) with forseti-security

File: deployment-templates/compute-engine/explain-instance.py (hunk 4 above)
Comment (jiyunyao) [reply deployment-templates/compute-engine/explain-instance.py]: I fixed this one. Can we open issues if there are other names needs to be changed? I don't want to accumulate too much fixes and improvements in this PR.

Comment (ahoying) [reply deployment-templates/compute-engine/explain-instance.py]: I'm fine with putting the variable rename changes in a future PR.

File: google/cloud/forseti/services/inventory/base/resources.py (hunk 5 above)
Comment (jiyunyao) [reply google/cloud/forseti/services/inventory/base/resources.py]: Fixed this one.

File: google/cloud/forseti/services/inventory/storage.py (hunk 6 above)
Comment (jiyunyao) [reply google/cloud/forseti/services/inventory/storage.py]: We don't have any case that there are rows in old_dict should be deleted. This is why we want delete information here.
There are 3 rows returns at most:
resource, iam_policy, gcs_policy
Right now the update is only used when adding error logs. 
Only resource has been updated when adding error logs.

File: scripts/iam_auto_deployment.sh (hunk 2 above)
Comment (jiyunyao) [reply scripts/iam_auto_deployment.sh]: Fixed

File: scripts/iam_auto_deployment.sh (hunk 3 above)
Comment (jiyunyao) [reply scripts/iam_auto_deployment.sh]: fixed

File: scripts/iam_auto_deployment.sh (hunk 7 above)
Comment (jiyunyao) [reply scripts/iam_auto_deployment.sh]: fixed

**Pull request merged**